src/scheduler/analysis.py

```diff
@@ -528,3 +640,269 @@ def _fix_system_version(system_version: str | None) -> str:
+            nonlocal run
+            run = False
+
+            if child_process is None:
+                return
+
+            child_process.join(Worker.SIGTERM_TIMEOUT)
+            if child_process.is_alive():
+                raise Worker.TimeoutError(Worker.SIGTERM_TIMEOUT)
+
+        signal.signal(signal.SIGTERM, _handle_sigterm)
+
+        while run:
+            try:
+                # We must have some non-infinite delay here to avoid blocking even after _handle_sigterm is called
+                task = self._in_queue.get(block=True, timeout=config.backend.block_delay)
+            except queue.Empty:
+                continue
+
+            entry = {}
+            try:
+                self._is_working.value = 1
+                logging.info(f'{self}: Begin {self._plugin.metadata.name} analysis on {task.scheduler_state.uid}')
+                start_time = time.time()
+
+                child_process = mp.Process(
+                    target=self._child_entrypoint,
+                    args=(self._plugin, task, send_conn),
+                )
+                child_process.start()
+                child_process.join(timeout=self._worker_config.timeout)
+                if not recv_conn.poll():
+                    raise Worker.TimeoutError(self._worker_config.timeout)
+
+                result = recv_conn.recv()
+
+                if isinstance(result, Exception):
+                    raise result
+
+                duration = time.time() - start_time
+
+                entry['analysis'] = result
+                logging.info(f'{self}: Finished {self._plugin.metadata.name} analysis on {task.scheduler_state.uid}')
+                if duration > 120:
+                    logging.info(
+                        f'Analysis {self._plugin.metadata.name} on {task.scheduler_state.uid} is slow: took {duration:.1f} seconds'
+                    )
+                self._update_duration_stats(duration)
+            except Worker.TimeoutError as err:
+                logging.critical(f'{self} timed out after {err.timeout} seconds.')
+                entry['timeout'] = (self._plugin.metadata.name, 'Analysis timed out')
+            except Exception as exc:  # pylint: disable=broad-except
+                # As tracebacks can't be pickled we just print the __exception_str__ that we set in the child
+                logging.critical(f'{self} got a exception during analysis: {exc}', exc_info=False)
+                logging.critical(exc.__exception_str__)
+                entry['exception'] = (self._plugin.metadata.name, 'Analysis threw an exception')
+            finally:
+                # Don't kill another process if it uses the same PID as our dead worker
+                if child_process.is_alive():
+                    child = psutil.Process(pid=child_process.pid)
+                    for grandchild in child.children(recursive=True):
+                        grandchild.kill()
+                    child.kill()
```

Comment:
could be moved into a function or maybe merged with something in `helperFunctions/process`